libdnf5/comps/comps_sack.cpp

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf5/comps/comps_sack.hpp"
+
+#include "comps_sack_impl.hpp"
+
+#include "libdnf5/comps/group/query.hpp"
+
+namespace libdnf5::comps {
+
+void CompsSack::Impl::load_config_excludes() {
+    const auto & main_config = base->get_config();
+
+    const auto & disable_excludes = main_config.get_disable_excludes_option().get_value();
+    if (std::find(disable_excludes.begin(), disable_excludes.end(), "*") != disable_excludes.end()) {
+        return;
+    }
+    if (std::find(disable_excludes.begin(), disable_excludes.end(), "main") != disable_excludes.end()) {
+        return;
+    }
+
+    for (const auto & name : main_config.get_excludeenvs_option().get_value()) {
+        EnvironmentQuery query(base, EnvironmentQuery::ExcludeFlags::IGNORE_EXCLUDES);
+        query.filter_environmentid(name, libdnf5::sack::QueryCmp::GLOB);
+        for (const auto & environment : query.list()) {
+            config_environment_excludes.insert(environment.get_environmentid());
+        }
+    }
+    for (const auto & name : main_config.get_excludegroups_option().get_value()) {
+        GroupQuery query(base, GroupQuery::ExcludeFlags::IGNORE_EXCLUDES);
+        query.filter_groupid(name, libdnf5::sack::QueryCmp::GLOB);
+        for (const auto & group : query.list()) {
+            config_group_excludes.insert(group.get_groupid());
+        }
+    }
+}
+
+const std::set<std::string> CompsSack::Impl::get_config_environment_excludes() {
+    return config_environment_excludes;
+}
+
+const std::set<std::string> CompsSack::Impl::get_config_group_excludes() {
+    return config_group_excludes;
+}
+
+const std::set<std::string> CompsSack::Impl::get_user_environment_excludes() {
+    return user_environment_excludes;
+}
+
+void CompsSack::Impl::add_user_environment_excludes(const std::set<std::string> & excludes) {
+    for (auto exclude : excludes) {
```

Comment:
We should avoid the unnecessary copies with `auto &` or `const auto &` if applicable. Also applies for other for loops in this file.